src/bitcoin/electrum/mod.rs

```diff
@@ -0,0 +1,217 @@
+        let server_hash = self.client.genesis_block().hash;
+        let wallet_hash = self.bdk_wallet.local_chain().genesis_hash();
+        if server_hash != *expected_hash || wallet_hash != *expected_hash {
+            return Err(ElectrumError::GenesisHashMismatch(
+                *expected_hash,
+                server_hash,
+                wallet_hash,
+            ));
+        }
+        Ok(())
+    }
+
+    pub fn client(&self) -> &client::Client {
+        &self.client
+    }
+
+    fn local_chain(&self) -> &LocalChain {
+        self.bdk_wallet.local_chain()
+    }
+
+    /// Get all coins stored in the wallet, taking into consideration only those unconfirmed
+    /// transactions that were seen in the last wallet sync.
+    pub fn wallet_coins(&self, outpoints: Option<&[OutPoint]>) -> HashMap<OutPoint, Coin> {
+        self.bdk_wallet.coins(outpoints, Some(self.sync_count))
+    }
+
+    /// Get the tip of the wallet's local chain.
+    pub fn wallet_tip(&self) -> BlockChainTip {
+        utils::tip_from_block_id(self.local_chain().tip().block_id())
+    }
+
+    /// Sync the wallet with the Electrum server. If there was any reorg since the last poll, this
+    /// returns the first common ancestor between the previous and the new chain.
+    pub fn sync_wallet(
+        &mut self,
+        receive_index: ChildNumber,
+        change_index: ChildNumber,
+    ) -> Result<Option<BlockChainTip>, ElectrumError> {
+        self.bdk_wallet.reveal_spks(receive_index, change_index);
+        let local_chain_tip = self.local_chain().tip();
+        log::debug!(
+            "local chain tip height before sync with electrum: {}",
+            local_chain_tip.block_id().height
+        );
+
+        const BATCH_SIZE: usize = 200;
+        // We'll only need to calculate fees of mempool transactions and this will be done separately from our graph
+        // so we don't need to fetch prev txouts. In any case, we'll already have these for our own transactions.
+        const FETCH_PREV_TXOUTS: bool = false;
+        const STOP_GAP: usize = 50;
+
+        let (chain_update, mut graph_update, keychain_update) = if local_chain_tip.height() > 0 {
+            log::info!("Performing sync.");
+            let mut request = SyncRequest::from_chain_tip(local_chain_tip.clone())
+                .cache_graph_txs(self.bdk_wallet.graph());
+
+            let all_spks: Vec<_> = self
+                .bdk_wallet
+                .index()
+                .inner() // we include lookahead SPKs
+                .all_spks()
+                .iter()
+                .map(|(_, script)| script.clone())
+                .collect();
+            request = request.chain_spks(all_spks);
+            log::debug!("num SPKs for sync: {}", request.spks.len());
+
+            let sync_result = self
+                .client
+                .sync_with_confirmation_time_height_anchor(request, BATCH_SIZE, FETCH_PREV_TXOUTS)
+                .map_err(ElectrumError::Client)?;
```

Comment:
here while simulating a reorg w/ a fulcrum server, right after invalidate some blocks (2), the poller will flood the server w/ sync request (w/o delay) as the request fail because local_height > chain_height:

```
2024-09-03T06:11:51.044431Z  INFO liana::bitcoin::electrum:135: Sync complete.

  2024-09-03T06:11:54.004698Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:54.042005Z  INFO liana::bitcoin::electrum:135: Sync complete.

  2024-09-03T06:11:57.001875Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.025950Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.044065Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.070370Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.084774Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.107061Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.114098Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.135826Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.148443Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.188261Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.203140Z  INFO liana::bitcoin::electrum:116: Performing sync.

  2024-09-03T06:11:57.241783Z ERROR liana::bitcoin::poller::looper:283: Error syncing wallet: 'Electrum client error: 'Electrum error: 'Electrum server error: {"code":1,"message":"Height 3007 is out of range"}'.'.'.

  2024-09-03T06:11:57.256700Z  INFO liana::bitcoin::electrum:116: Performing sync.
```
and stop flooding when chain_height => local_height

I don't think this can happen in production as the chain_height should never be < local_height during a reorg, but maybe we should keep the delay between 2 sync request even if the request fail (in the caller of the method, not here)?

Reply:
Thanks for finding this issue. I think a short delay before retrying in the poller would make sense.

We might also have different retry behaviour for different error types but I'm not sure if there are standard Electrum errors so this might be tricky.

Reply:
i think this can be in a follow up.
at the moment i've test w/ electrs/fulcrum/electrumx, they both have different behaviour...

Reply:
OK thanks for checking that. I think the sleep between retries would still make sense to add now.

Reply:
Added sleep for 2 seconds.